packages/nimble-components/src/table/specs/table-row-grouping-hld.md

```diff
@@ -0,0 +1,298 @@
+    }
+
+    ...
+}
+
+const tableGroupRowElement = TableGroupRow.compose({
+    baseName: 'table-group-row',
+    template: html<TableGroupRow>`
+    <template
+        role="rowgroup"
+        style="--ni-private-row-group-indent-level: ${x => x.nestingLevel};"
+    >
+        <nimble-button
+            appearance="ghost"
+            content-hidden
+            @click=${x => x.onExpandGroupToggle()}>
+            ${when(x => x.isExpanded, html`
+                <nimble-icon-arrow-expander-down slot="start"></nimble-icon-arrow-expander-down>
+            `)}
+            ${when(x => !x.isExpanded, html`
+                <nimble-icon-arrow-expander-right slot="start"></nimble-icon-arrow-expander-right>
+            `)}
+        </nimble-button>
+        <${x => x.rowGroupState.groupRowHeaderViewTag}
+            :rowGroupValue="${x => x.groupRowValue}"
+            :columnConfig="${x => x.columnConfig}"
+            >
+        </{x => x.rowGroupState.groupRowHeaderViewTag}> (${x => x.leafItemCount})
+    </template>
+    styles: /* styling */
+});
+DesignSystem.getOrCreate().withPrefix('nimble').register(tableGroupRowElement());
+```
+
+Grouped rows are provided alongside ungrouped rows through the TanStack function `getRowModel().rows`, which is already used for retrieving the rows to render. We can then provide all of the necessary state to render a `TableGroupRow` alongside the existing state in `TableRowState`. So to render the `TableGroupRows`, the template can make a simple change similar to this:
+
+Table template.ts
+
+```ts
+    ...
+    ${repeat(x => x.virtualizer.visibleItems, html<VirtualItem, Table>`
+        ${when((x, c) => (c.parent as Table).tableData[x.index]?.isGrouped, html<VirtualItem, Table>`
+            <${DesignSystem.tagFor(TableGroupRow)}
+                :groupRowValue="${(x, c) => c.parent.tableData[x.index]?.groupRowValue}"
+                ...
+                @on-row-group-expand-toggle="${(x, c) => c.parent.toggleRowExpanded(x.index)}"
+                >
+            </${DesignSystem.tagFor(TableGroupRow)}>
+        `)}
+        ${when((x, c) => !(c.parent as Table).tableData[x.index]?.isGrouped, html<VirtualItem, Table>`
+            <${DesignSystem.tagFor(TableRow)}
+    ...
+```
+
+The one change necessary to support this properly is to change the virtualizer to use the current length of `getRowModel.rows()` for the `count` instead of the [current length of the table's data](https://github.com/ni/nimble/blob/f60dedf600147b16916f876b749a4e019c2a6308/packages/nimble-components/src/table/models/virtualizer.ts#L82) (see [prototype](https://github.com/ni/nimble/blob/5cd05ab733b00ebb33e44d53c7be30c980fa3398/packages/nimble-components/src/table/models/virtualizer.ts#L87)).
+
+### Row interaction summary
+
+-   Group rows will perform an expand/collapse operation when clicked anywhere on the row.
+-   The expand/collapse button will still have button hover/click visual state
```

Comment:
Is this true? From my understanding based on what you said yesterday is that the "button" would not have hover/click visual state -- just the row would have a hover state.

Reply:
You are correct. I hadn't updated the HLD with the new guidance. Updated now.